File: TKDGTT/Tuan 3/BT3.LamThiHang.TranThiThuHang.VuThiMinhHang.TranManhHung/Bai3/max_min.cpp
#include<stdio.h>
#include<conio.h>

void nhap_mang(int a[], int n){
	int i;
	for(i=0;i<n;i++){
		printf("Nhap a[%d] =  ",i+1);
		scanf("%d", &a[i]);
		}
	}
	
void in_mang(int a[], int n){
	int i;
	for(i=0;i<n;i++)
		printf("%4d", a[i]);
	printf("\n");
	}
	
int max_day(int x, int y){
	if(x>y) return x;
	return y;
	}
	
int min_day(int x, int y){
	if(x<y)  return x;
	return y;
	}

void max_min(int a[], int l, int r,int &max, int &min){
	int max1,max2,min1,min2;
	if(r-l<=1){
         max = max_day(a[l],a[r]);
         min = min_day(a[l],a[r]);      
	}
	else{
	    max_min(a,l,(l+r)/2,max1,min1);
		max_min(a,(l+r)/2,r,max2,min2);
		max = max_day(max1,max2);
		min = min_day(min1,min2);
	}
}

int main(){
	int a[50];
	int n,max,min;
	printf("Nhap vao do dai day :  ");
	scanf("%d",&n);
	nhap_mang(a,n);
	printf("Day so vua nhap la : \n");
	in_mang(a,n);
	max_min(a,0,n-1,max,min);
	printf("Phan tu lon nhat va nho nhat la %d va %d.\n",max,min);
	getch();
	return 0;
	}

File: TKDGTT/Bai Tap/Bai Tap Giua Ky/TL.Nhom15. Phuong Phap Chia De Tri/code/MaxMin.java
import java.util.*;

public class MaxMin {

    static int[] arr;
    int max, min;

    public MaxMin(int n) {
        arr = new int[n];
        Random rd = new Random();
        for (int i = 0; i < n; i++) {
        	arr[i]=rd.nextInt(100);
        }
    }

    public int getMax() {
        return max;
    }

    public int getMin() {
        return min;
    }

    public void print() {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + "|");
        }
        System.out.println();
    }
    
    private void dacMaxMini() {
        dacMaxMin(arr,0, arr.length - 1);
		
	}
    public void dacMaxMin(int[] array, int left, int right) {
        int max1, min1, mid, max2, min2;
        if(right-left<=1){
        	if(array[left]<array[right]){
        		max = array[right];
                min = array[left];
            }
        	else {
                max = array[left];
                min = array[right];
            }
        }
        else{
        	mid=(left+right)/2;
        	dacMaxMin(array,left,mid);
        	max1=max;
        	min1=min;
        	dacMaxMin(array,mid+1,right);
        	max2=max;
        	min2=min;
        	if (max2 < max1) {
                max = max1;
            }
            if (min2 > min1) {
                min = min1;
            }
        }
        
    }

    public static void main(String[] a) {
        MaxMin m = new MaxMin(15);
        System.out.println("In Mang:");
        m.print();
        m.dacMaxMini();
        System.out.println("Gia tri lon nhat trong mang : " + m.getMax());
        System.out.println("Gia tri nho nhat trong mang : " + m.getMin());
    }

	

}
